### DISCUSSION TITLE
Extend the limits of Difficulty Adjust

### DISCUSSION THREAD

**Discussion by Stoppedpuma** [ppy/osu#27526] [Ideas]
Moved from #27523.

Idea is to extend the limits from 11 to 12.5 (besides CS since it becomes negative in size at 12.15) to match [McOsu](https://github.com/McKay42/McOsu). Is 12.5 still a possibility or would it be 12 at most to be consistent between all options?

**Comment by Rekunan** [comment]
I think it would be ideal to preserve consistency. Maybe to unlock 12.5 we could have CS stay the same size after CS12.1? I'd have to check to see how McOsu handles it.

**Reply by Stoppedpuma** [reply]
>12.5 we could have CS stay the same size after CS12.1

I'm all for consistency but this is just a really bad way of handling this imo, if we're doing this much for consistency then it's basically the same boat as adding -10 to all values as well just because AR has it. I think it'd be a different story if it was playable by a large group of people but I doubt anyone would even attempt to play this way? I think I'd prefer losing a small bit of consistency to allow most of the options to have a higher maximum value that players have been using for years and not have to cut them short because of an issue with one of them? All depends on what the dev team says but even bumping them all to 12 would be an amazing change in my eyes.

**Reply by Rekunan** [reply]
> this is just a really bad way of handling this imo

Yeah, thinking about it now, it sounds dumb to have CS12.1 be the same as CS12.5 I checked mcosu, and it seems they just capped it at CS12.14, so prob just all 12 is fine.

> adding -10 to all values

That's a good point, yeah, that should probably happen sometime as well.

**Reply by Stoppedpuma** [reply]
Just to make sure, ideally in your eyes -10 minimum and 12 maximum across all options? This sounds fine to me as it's still an improvement.

**Reply by Rekunan** [reply]
Yes, tho I do wanna see how -10 affects the other stats first.

**Reply by Stoppedpuma** [reply]
[CS -10](https://puma.s-ul.eu/sLabf5j5)
[HP -10](https://puma.s-ul.eu/Aok75T6x)
[AR -10](https://puma.s-ul.eu/hUV60rt9)
[CS HP AR -10](https://puma.s-ul.eu/CqVB7GEz)

**Reply by Rekunan** [reply]
For HP it'd be better to show how it decreases for missing or whatever, similar thing for OD (which u didn't include).

**Reply by Stoppedpuma** [reply]
> For HP it'd be better to show how it decreases for missing

This is one of the issues with making everything consistent, HP will not drain starting at -1.8. Not only is basically no one going to use values like these for HP and OD, both require changes to work at all in the case of OD breaking completely with spinners in negative values.

When it runs into things like HP becoming no fail at a point it makes more sense to me to have a bit of inconsistency to be able to push values that can be pushed higher to be higher and vise versa.

[HP -10 Gameplay](https://puma.s-ul.eu/a4eCtLRy)
[OD -10 Gameplay](https://puma.s-ul.eu/QImtFfuf)

**Reply by McKay42** [reply]
> I think it would be ideal to preserve consistency. Maybe to unlock 12.5 we could have CS stay the same size after CS12.1? I'd have to check to see how McOsu handles it.

> Yeah, thinking about it now, it sounds dumb to have CS12.1 be the same as CS12.5 I checked mcosu, and it seems they just capped it at CS12.14, so prob just all 12 is fine.

Correct, it's hard clamped to `12.1429` specifically. CS 12 is already just 0.64 osu!pixels so higher values don't make much sense for gameplay anyway.

**Reply by Rekunan** [reply]
> issues with making everything consistent

If the work to counteract the issues is too much then yeah maybe, but if possible, I'd much prefer being able to keep consistency, as it means that new users won't be confused why they can't go negative for other values, and waste time asking why.

**Reply by Stoppedpuma** [reply]
> I'd much prefer being able to keep consistency

I'm sure everyone wants to keep things consistent, I like to keep things consistent as well as long as it doesn't limit or break things. Currently the problems that are making me against consistency in this case are "stupid values" such as negative OD and HP. Negative HP becomes Nofail at a point, and negative OD breaks spinners completely. Personally I like how HP is now and I would rather not refactor spinners for a value that I can't see anyone using.

Those are the main things that are stopping me from completely agreeing with consistency in this situation. I would prefer being able to push values that can be pushed higher to be so, the same applies to lowering values like CS and AR where I can see Easy players having the time of their lives.

**Comment by Givikap120** [comment]
I see CS above 11 as unnecessary, because it's already barely playable. CS12 being less than 1px on the screen is a sign that it's not playable.
AROD above 11 is useful for HT, especially 0.5x, so don't see why it shouldn't exist.

**Reply by Stoppedpuma** [reply]
> CS above 11 as unnecessary, because it already barely playable

Fully agree here, only reason this was ever going to exist in my eyes was for consistency which I don't even think should be a big priority for difficulty adjust.

> AROD above 11 is useful for HT

These are probably the main ones people will use and are fully playable even without HT, I would like to push these to higher values such as 12.5 for AR, OD, and HP. Currently the issue is some would rather consistency more than making limits to be attribute specific. Limiting all values to a certain number just because one is capped is a bad way of doing things for a mod like difficulty adjust in my eyes, same applies to adding negative values to every attribute just because it makes sense with one or two of them. I think difficulty adjust should be able to push limits far but not to a point where it becomes unplayable or plain stupid.

**Comment by Stoppedpuma** [comment]
Just to put some numbers out to push this forward as I would like to implement this in a way which fits both values that are playable and somewhat the want for consistency how do these values sound?

CS -12.5 - 11.5
HP 12.5
OD 12.5
AR -12.5 - 12.5

@Rekunan are these values close enough for you concerning consistency? CS 12 is just not something I see feasible with this change, especially when the only CS 11.5+ FC I've ever seen is a map as simple as the one you used in your example of a CS 11.9 FC.

**Reply by Givikap120** [reply]
negative 12.5 CS bro what
CS0 is already on the edge of being useful, that's why peppy limited it to CS2 in the first place, do you want to have circles with the size of the screen or what
HP12.5 is absolutely useless, just use SD at this point
AR is already limited at -10 what is more than enough

adding useless things for the sake of consistency sounds like very bad idea to me

**Reply by Stoppedpuma** [reply]
> negative 12.5 CS bro what + AR is already limited at -10

Please read the rest of the discussion, I gave a reason for this. To summarise, I don't see a reason why this cant be a thing when it makes things harder to read, easy players exist and play maps with extremely low CS and AR all the time.

> that's why peppy limited it to CS2 in the first place

You do realise CS 0 is standarised as the minimum in lazer right?

> HP12.5 is absolutely useless, just use SD at this point + OD12.5 is humanly impossible to acc

At this rate I don't even think you know what you're talking about and have never used these values. HP 12.5 can miss 4 times in a row from full HP which immediately invalidates your points of useless and to use SD instead. OD 12.5 is not "humanly impossible" to play accurately? If I'm consistently able to get 85%+ on 5+ star maps even with lazers 10 ms audio buffer how is that not playable? Not to mention you literally asked for higher values above "AROD above 11 is useful for HT, especially 0.5x, so don't see why it shouldn't exist.".

**Reply by Rekunan** [reply]
I would still like to see CS12, it being barely playable now doesn't mean it won't be playable in the future. The same guy who did that CS11.9 FC said he was rly close to a CS12 FC.

**Reply by Stoppedpuma** [reply]
My problem still with CS 12 is I see it as a "stupid value". While yes it might be possible to fc a map with circles in the same 4 spots, I don't see anyone playing ranked maps above maybe 2-3 stars. My other concern with this is consistency, CS 12 matches in terms of numbers but not at all in difficulty? I don't think anyone else will think that difficulty consistency matters because even I think it's a bit far fetched to match based off that but by the time we're adding CS 12 we might as well be adding 13+ to other values to make difficulty feel consistent.

I chose 12.5 for all values except CS because they are playable in my eyes (except AR but HT helps a bit) and are mostly consistent all ending with .5. I'll wait a while to see what others say but what the majority agrees on is what I'll be PRing since I'd like to see these values updated sometime in the next few releases.

**Reply by Givikap120** [reply]
Easy players don't play with CS0 and AR-10. The lowest you can see in normal plays is AR0 with EZHT. And the point was - why TF you should increase limit from AR-10 to AR-12.5: you're changing nice number to a bad a stupid one for no reason. And don't even talk about consistency here because there are no other -12.5 values.

Peppy limited CS to 2 in stable editor so mappers won't use CS<2 and CS>7 because it was considered "unplayable" by peppy. You can still put this AR with editing map file directly.

Ok, tell me 1 case where player actually wants to set HP to 12.5. I can't even imagine why anyone should use HP10. Like in stable HP10 is completely useless and unplayable, in lazer it's playable, but still - why would player overwrite HP value to 10, let alone 12.5.

OD - i edited that out because i forgot about HT existing when writing the comment. In NM OD12.5 is humanly impossible, your subjective "i can get 85 acc" can't argue with mathematical facts. The lower UR ever achievable is 35.5, and this is SUPER edgecase. And guess what? With OD12.5 this UR is equivalent to 90% acc.
Also, 85 acc is straight up lie or manipulation with you probably using smth like CL mod. Because i doubt you can achieve 45UR to get 85 acc on a map.

**Reply by Stoppedpuma** [reply]
> Easy players don't play with CS0 and AR-10

This fully confirms to me that you have no idea what in the world you are talking about let alone apparently the ability to read the rest of the discussion. Lets use an example of the term "negative infinity ar" where do you think it came from and where were people playing it on? The client with values I'm trying to have matched, McOsu.

> Peppy limited CS to 2 in stable editor

What is your point here? No where in this discussion is anyone worrying about stable. Stable is not the way forward if you haven't realised.

> tell me 1 case where player actually wants to set HP to 12.5

Tell me one case there wouldn't be a player who would want to minimise their misses as a challenge and seeing raising HP as an option? I could say things like #27198 but that's overall misses but still an example of someone wanting to minimise misses.

> In NM OD12.5 is humanly impossible + With OD12.5 this UR is equivalent to 90% acc.

Humanly impossible yet 90% accuracy? Dude what the fuck are you talking about? 90% acc is an A rank and if you see that as humanly impossible then you probably say some really stupid shit like that a 10 star fc is humanly impossible because the players in 2015 couldn't do it. If you can't even try the values that you're calling impossible then you probably shouldn't be discussing **values in a mod which are able to be changed to extreme amounts.** I don't want you to not have place in this discussion since all opinions are valid concerning this but if you're just going to point fingers constantly and go on a rant about things you seem to have no idea about then they aren't desirable to moving this conversation forward.

**Reply by Stoppedpuma** [reply]
@Rekunan to get some more numbers on the table what are some ranges you would like to see?

**Reply by Rekunan** [reply]
I think the other ranges u suggested are fine.

**Reply by Stoppedpuma** [reply]
A few questions I have for you:

Are you fine with CS 11.5 being the max for now?
Should negative CS be an option?
Would you prefer values ending with .5 or being a whole number?

**Reply by Rekunan** [reply]
I would much prefer CS12, since I don't really see any valid reasons against it.
For the other 2 questions, I have no preference.

**Reply by Stoppedpuma** [reply]
@Rekunan @Givikap120 How about just making everything cap at 12?

CS 12
HP 12
OD 12
AR -12 - 12

**Reply by Rekunan** [reply]
I can accept this, yes.

**Reply by Stoppedpuma** [reply]
If no one has any objections about limiting everything to 12 over the next few days then I'll submit a PR using these values.

**Comment by Stoppedpuma** [comment]
Since we can't seem to agree on values how about letting the user decide? This would cap all values at the absolute limit and automatically lower the value if someone goes over it, example: AR 13 is 0 ms, if a user enters AR 14 the box would change to a shade of red and shake similarly to notelock then return to it's normal colour and set the value to 13 since it's the literal max when the user presses enter. This option could be hidden behind right clicking the slider bar to reveal a text input.

![1711086832](https://github.com/ppy/osu/assets/58333920/195c4cbf-6f20-4ab2-a586-ab7095f094ad)
![1711086838](https://github.com/ppy/osu/assets/58333920/f435735b-28b6-42be-93f6-61e1aa90cb5c)

This would definitely need additional information to be shown when hovering over values similarly to how AR shows the total time in ms. OD example: Great: x ms Good: x ms Meh: x ms.

This will also leave the sliders current caps where they are at the moment since they are good caps for a large chunk of the player base and the text box will allow those select few who want to go further to do so.

**Reply by Rekunan** [reply]
I disagree with having it hidden like this, this sounds like poor UX. Much better to just have the limit be obvious.

**Reply by Stoppedpuma** [reply]
Would either of these work? It doesn't match the rest of the UI that much but I agree showing the value would be beneficial.

11.8 (users desired value) / 13.0 (maximum) 11.8 / 13.0 
-x (minimum) [users desired value] 13.0 (maximum)

Regarding being hidden and a few reasons this might not be a good idea at the moment:

I've thought about this a bit more and none of the other mods have options like this, both in being hidden behind a right click (I'm pretty sure you meant hidden being the maximum number but I'm going to apply it here too) and in how it's shown since it's normally just an input box, not a slider with an input box beside it. I think waiting on this idea of how to handle DA might be a better way to go about.

I think just adjusting slider values should be enough for now since there's a lot to consider when giving users this much customisability, if we were to jump into this idea things like negative ar can be brought to values similar to the term "negative infinity ar"  which leads to massive performance issues and (I think) has already been decided to be implemented as an entirely seperate mod #16732. This is probably not a great idea at the moment when there's already so much discussion happening centred around mods.

**Reply by Rekunan** [reply]
I'm having trouble reading this message, so lmk if I interpreted anything wrong.

I think the best choice is just using what we already have, so just showing the value as is, and then not letting them go beyond the limits.

**Reply by Stoppedpuma** [reply]
Yeah that's what I meant, to simplify the above it's basically just saying it's out of place in the UI and needs a ton of consideration before we allow users to have that much customisation.

**Comment by Givikap120** [comment]
Having all-time visible number boxes is actually a very good idea and can potentially save-up a lot of time, especially on mobile devices

**Reply by Stoppedpuma** [reply]
Have to agree here, especially on mobile. This should probably be a global interaction if anything. Maybe for the time being right click on desktop and hold on mobile to reveal an input box? Interaction with the slider bar is the first idea on my mind at the moment to avoid redesigning menus or menu clutter.

I've opened a discussion for this at #27702.
